In [1]:
# importing the necessary libraries for data analysis
import pandas as pd
import numpy as np
import glob

In [2]:
# Combine all city data into one big table
city_files = glob.glob("Dataset/*.csv")

all_cities_data = []


for file_name in city_files:
    
    city_df = pd.read_csv(file_name)
    all_cities_data.append(city_df)
    print(f"Loaded: {file_name}")
    
    
combined_data = pd.concat(all_cities_data, ignore_index=True)


#Save the combined data to a new CSV file
#combined_data.to_csv("all_cities_combined.csv", index=False)


print(f"SUCCESS: Combined {len(city_files)} city files into one table with {len(combined_data)} total rows")

Loaded: Dataset\PRSA_Data_Changping_20130301-20170228.csv
Loaded: Dataset\PRSA_Data_Dongsi_20130301-20170228.csv
Loaded: Dataset\PRSA_Data_Guanyuan_20130301-20170228.csv
Loaded: Dataset\PRSA_Data_Shunyi_20130301-20170228.csv
SUCCESS: Combined 4 city files into one table with 140256 total rows


In [3]:
## preview of combined data

#df= pd.read_csv('all_cities_combined.csv')
#df

combined_data

,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station
0,1,2013,3,1,0,3.0,6.0,13.0,7.0,300.0,85.0,-2.3,1020.8,-19.7,0.0,E,0.5,Changping
1,2,2013,3,1,1,3.0,3.0,6.0,6.0,300.0,85.0,-2.5,1021.3,-19.0,0.0,ENE,0.7,Changping
2,3,2013,3,1,2,3.0,3.0,22.0,13.0,400.0,74.0,-3.0,1021.3,-19.9,0.0,ENE,0.2,Changping
3,4,2013,3,1,3,3.0,6.0,12.0,8.0,300.0,81.0,-3.6,1021.8,-19.1,0.0,NNE,1.0,Changping
4,5,2013,3,1,4,3.0,3.0,14.0,8.0,300.0,81.0,-3.5,1022.3,-19.4,0.0,N,2.1,Changping
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
140251,35060,2017,2,28,19,27.0,72.0,8.0,92.0,800.0,16.0,10.3,1014.2,-12.4,0.0,W,1.8,Shunyi
140252,35061,2017,2,28,20,47.0,55.0,17.0,86.0,1100.0,19.0,9.8,1014.5,-9.9,0.0,NW,1.5,Shunyi
140253,35062,2017,2,28,21,18.0,28.0,4.0,30.0,500.0,64.0,9.1,1014.6,-12.7,0.0,NE,1.7,Shunyi
140254,35063,2017,2,28,22,18.0,20.0,9.0,33.0,500.0,59.0,7.1,1015.2,-13.2,0.0,WNW,1.8,Shunyi


**Basic Data Understanding:**

To get familiar with  dataset, I started by examining the columns and their corresponding data types. I also reviewed the descriptive statistics and checked for any missing values.

In [4]:
combined_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140256 entries, 0 to 140255
Data columns (total 18 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   No       140256 non-null  int64  
 1   year     140256 non-null  int64  
 2   month    140256 non-null  int64  
 3   day      140256 non-null  int64  
 4   hour     140256 non-null  int64  
 5   PM2.5    137203 non-null  float64
 6   PM10     138144 non-null  float64
 7   SO2      137195 non-null  float64
 8   NO2      135964 non-null  float64
 9   CO       131607 non-null  float64
 10  O3       136326 non-null  float64
 11  TEMP     140112 non-null  float64
 12  PRES     140115 non-null  float64
 13  DEWP     140109 non-null  float64
 14  RAIN     140114 non-null  float64
 15  wd       139474 non-null  object 
 16  WSPM     140141 non-null  float64
 17  station  140256 non-null  object 
dtypes: float64(11), int64(5), object(2)
memory usage: 19.3+ MB


In [5]:
combined_data.head()

,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station
0,1,2013,3,1,0,3.0,6.0,13.0,7.0,300.0,85.0,-2.3,1020.8,-19.7,0.0,E,0.5,Changping
1,2,2013,3,1,1,3.0,3.0,6.0,6.0,300.0,85.0,-2.5,1021.3,-19.0,0.0,ENE,0.7,Changping
2,3,2013,3,1,2,3.0,3.0,22.0,13.0,400.0,74.0,-3.0,1021.3,-19.9,0.0,ENE,0.2,Changping
3,4,2013,3,1,3,3.0,6.0,12.0,8.0,300.0,81.0,-3.6,1021.8,-19.1,0.0,NNE,1.0,Changping
4,5,2013,3,1,4,3.0,3.0,14.0,8.0,300.0,81.0,-3.5,1022.3,-19.4,0.0,N,2.1,Changping


In [6]:
combined_data.columns

Index(['No', 'year', 'month', 'day', 'hour', 'PM2.5', 'PM10', 'SO2', 'NO2',
       'CO', 'O3', 'TEMP', 'PRES', 'DEWP', 'RAIN', 'wd', 'WSPM', 'station'],
      dtype='object')

**Description of dataset columns:**



Year – The calendar year when the air quality measurement was taken.

Month – The month of the measurement, useful for seasonal or monthly trend analysis.

Day – The day of the month when the reading was recorded.

Hour – The hour of the day (0–23), allowing analysis of hourly pollution patterns.

PM2.5 – Fine particulate matter smaller than 2.5 micrometers. These particles penetrate deep into the lungs and are linked to serious health risks.

PM10 – Particulate matter smaller than 10 micrometers, including dust, smoke, and coarse particles.

SO2 – Sulfur dioxide, a pollutant produced mainly from burning coal or oil; contributes to respiratory irritation and acid rain.

NO2 – Nitrogen dioxide, a toxic gas from combustion sources like vehicles and industry; contributes to smog and respiratory issues.

CO – Carbon monoxide, a colorless gas from incomplete combustion; harmful at high concentrations.

O3 – Ground-level ozone, formed by chemical reactions in sunlight; a major component of smog and harmful to lungs.

TEMP – Ambient air temperature (°C) at the time of measurement.

PRES – Atmospheric pressure (hPa), useful for understanding weather conditions influencing pollution.

DEWP – Dew point temperature (°C), indicating moisture content in the air.

RAIN – Rainfall amount, typically in millimeters; rainfall can reduce particulate pollution.

wd – Wind direction, indicating the compass direction from which the wind is blowing (e.g., NW, NNE).

WSPM – Wind speed (m/s), which affects pollutant dispersion.

station 


In [7]:
combined_data.shape

(140256, 18)

# **Inference :**


---


The dataset contains 14,0256 records with 18 columns, and while the station and Date columns are fully complete, several pollutant measurements show missing data. Key pollutants such as PM2.5, PM10, NH3, Toluene, and Xylene have a significant number of missing values, with Xylene having the least completeness.

In [8]:
City = combined_data['station'].value_counts()
print(f'Total number of cities in the dataset : {len(City)}')
City

Total number of cities in the dataset : 4


station
Changping    35064
Dongsi       35064
Guanyuan     35064
Shunyi       35064
Name: count, dtype: int64

In [9]:
combined_data['station'].unique()

array(['Changping', 'Dongsi', 'Guanyuan', 'Shunyi'], dtype=object)

In [10]:
print(f"There are  {combined_data.duplicated().sum()} duplicates in the dataset")

There are  0 duplicates in the dataset



### **Dataset Structure**

The dataset contains **140,256 rows and 18 columns**. It includes hourly air‑quality pollutant measurements, meteorological variables (TEMP, PRES, DEWP, RAIN, wind direction and speed), and station names that represent the monitoring locations. The date information is stored across **year, month, day, and hour** columns.

There are two main types of columns in the dataset:

-   **Categorical columns:** _station_, _wd_, and the combined _date_ fields when treated as a categorical timestamp (3 columns)
    
-   **Numerical columns:** All pollutant concentrations (PM2.5, PM10, SO2, NO2, CO, O3) and meteorological variables (TEMP, PRES, DEWP, RAIN, WSPM), along with the index-like _No_ column (13 columns)



### **Data Completeness**

The dataset is largely complete, with **no missing values** in the columns: **year, month, day, hour, station, No**

The pollutant and weather variables show varying levels of missing data:

-   **Least missing:** _DEWP (0.10%)_, _TEMP (0.10%)_, _RAIN (0.10%)_, _PRES (0.10%)_, _WSPM (0.08%)_, _wd (0.56%)_
    
-   **Moderate missing:** _PM10 (1.50%)_, _PM2.5 (2.18%)_, _SO2 (2.18%)_, _O3 (2.80%)_, _NO2 (3.06%)_
    
-   **Most missing:** **CO (6.17%)** — the variable with the highest proportion of missing values in the dataset


In [11]:
#checking Missing values
def missing_values_table(df):
    mis_val = df.isnull().sum()

  
    mis_val_percent = 100 * mis_val / len(df)

    mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)

    mis_val_table = mis_val_table.rename(columns={0: 'Missing Values', 1: '% of Total Values'})

    mis_val_table = mis_val_table.sort_values('% of Total Values', ascending=False)

    return mis_val_table

missing_values = missing_values_table(combined_data)
display(missing_values.style.background_gradient(cmap='Greens'))

,Missing Values,% of Total Values
CO,8649,6.166581
NO2,4292,3.060119
O3,3930,2.802019
SO2,3061,2.182438
PM2.5,3053,2.176734
PM10,2112,1.505818
wd,782,0.557552
DEWP,147,0.104808
TEMP,144,0.102669
RAIN,142,0.101243
